pkg/visor/init.go

```diff
@@ -34,54 +35,132 @@ import (
 	"github.com/skycoin/skywire/pkg/util/updater"
 	"github.com/skycoin/skywire/pkg/visor/hypervisorconfig"
 	"github.com/skycoin/skywire/pkg/visor/visorconfig"
+	vinit "github.com/skycoin/skywire/pkg/visor/visorinit"
 )
 
-type initFunc func(v *Visor) bool
-
-func initStack() []initFunc {
-	return []initFunc{
-		initUpdater,
-		initEventBroadcaster,
-		initAddressResolver,
-		initDiscovery,
-		initSNet,
-		initDmsgpty,
-		initTransport,
-		initRouter,
-		initLauncher,
-		initCLI,
-		initHypervisors,
-		initUptimeTracker,
-		initTrustedVisors,
-		initHypervisor,
-	}
+type visorCtxKey int
+
+const visorKey visorCtxKey = iota
+
+var up, ebc, ar, disc, sn, pty, tr, rt, launch, cli, hvs, ut, trv, hv vinit.Module
+
+// visor module to group all visor dependencies
+var vis vinit.Module
+
+// todo: return error from init functions with meaningful values instead of bool
+// todo: consider refactoring reporting system
+
+func registerModules(logger *logging.MasterLogger) {
+	// procedure:
+	// look at init function for module x. if init function makes use of
+	// field v.y for some module y of visor v, then y depends on x.
+	// if init function for x passes v itself, it's fucked up. All deps should be added, but
+	// better yet this should be refactored to pass the actual dep
+	up = vinit.MakeModule("updater", initUpdater, logger)
+	ebc = vinit.MakeModule("event_broadcaster", initEventBroadcaster, logger)
+	ar = vinit.MakeModule("address_resolver", initAddressResolver, logger)
+	disc = vinit.MakeModule("discovery", initDiscovery, logger)
+	sn = vinit.MakeModule("snet", initSNet, logger, &ar, &disc, &ebc)
+	pty = vinit.MakeModule("dmsg_pty", initDmsgpty, logger, &sn)
+	tr = vinit.MakeModule("transport", initTransport, logger, &sn, &ebc)
+	rt = vinit.MakeModule("router", initRouter, logger, &tr, &sn)
+	launch = vinit.MakeModule("launcher", initLauncher, logger, &ebc, &disc, &sn, &tr, &rt)
+	cli = vinit.MakeModule("cli", initCLI, logger)
+	hvs = vinit.MakeModule("hypervisors", initHypervisors, logger, &sn)
+	ut = vinit.MakeModule("uptime_tracker", initUptimeTracker, logger)
+	trv = vinit.MakeModule("trusted_visors", initTrustedVisors, logger, &tr)
+	vis = vinit.MakeModule("visor", vinit.DoNothing, logger, &up, &ebc, &ar, &disc, &sn, &pty,
+		&tr, &rt, &launch, &cli, &hvs, &ut, &trv)
+	hv = vinit.MakeModule("hypervisor", initHypervisor, logger, &vis)
 }
 
-func initUpdater(v *Visor) bool {
-	report := v.makeReporter("updater")
+// -------------------------------------------------------------------
 
+func initUpdater(ctx context.Context, log *logging.Logger) error {
+	v, err := getVisor(ctx)
```

Comment:
These functions require visor, and ideally need visor argument. But it's not possible because that would result in circular import: the modules package will depend back on visor.

I initially used wrapping, but ended up with a much more cumbersome method that what it is currently:
https://github.com/skycoin/skywire/blob/92e84034eea0081c29d0ad2b2fe6e543856345a9/pkg/visor/init.go

Extra indentation and multiple context arguments were just too much, so I just made it dead-simple. But now that I think about it there seems to be a better way of doing it than I did the first time. I'll try to move it somewhere to avoid code repetition, thanks.